janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import org.apache.tinkerpop.gremlin.process.traversal.*;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.DefaultGraphTraversal;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.FilterStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.IdStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.janusgraph.core.Cardinality;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.core.RelationType;
+import org.janusgraph.graphdb.internal.InternalRelationType;
+import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+import org.janusgraph.graphdb.types.CompositeIndexType;
+import org.janusgraph.graphdb.types.IndexType;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.janusgraph.graphdb.types.system.ImplicitKey.ADJACENT_ID;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class AdjacentVertexHasUniquePropertyOptimizerStrategy
+    extends AdjacentVertexOptimizerStrategy<HasStep<?>> {
+
+    private static final AdjacentVertexHasUniquePropertyOptimizerStrategy INSTANCE =
+        new AdjacentVertexHasUniquePropertyOptimizerStrategy();
+
+    private AdjacentVertexHasUniquePropertyOptimizerStrategy() {}
+
+    public static AdjacentVertexHasUniquePropertyOptimizerStrategy instance() { return INSTANCE; }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        TraversalHelper.getStepsOfClass(HasStep.class, traversal)
```

Comment:
I am not sure here, but maybe you want to skip this optimization if `TraversalHelper.onGraphComputer(traversal)`. TBH I didn't even know we could create nested traversal, thus I don't know what impact it will have on OLAP traversal.

Reply:
You're right. I never did any experiments with OLAP traversals so it's probably wise to not tinker (pun intended) with them before gaining a deeper knowledge.

Reply:
nice pun!